R/grade_this_learnr.R

```diff
@@ -0,0 +1,286 @@
+#' @param last_value The last value from evaluating the exercise.
+#' @param ... Extra arguments supplied by learnr
+#'
+#' @return An R list which contains several fields indicating the result of the
+#'   check.
+#' @export
+#' @seealso [gradethis_setup()]
+#'
+#' @examples
+#' \dontrun{gradethis_demo()}
+grade_this_learnr <- function(label = NULL,
+                         solution_code = NULL,
+                         user_code = NULL,
+                         check_code = NULL,
+                         envir_result = NULL,
+                         evaluate_result = NULL,
+                         envir_prep = NULL,
+                         last_value = NULL,
+                         ...) {
+  # Call this function in such a way that it can use other gradethis internals when called by learnr
+  # (i.e., make tutorial_options(exercise.checker = gradethis::grade_this_learnr) always work)
+  utils::getFromNamespace("grade_this_learnr_", "gradethis")(
+    label = label,
+    solution_code = solution_code,
+    user_code = user_code,
+    check_code = check_code,
+    envir_result = envir_result,
+    evaluate_result = evaluate_result,
+    envir_prep = envir_prep,
+    last_value = last_value,
+    ...
+  )
+}
+
+
+grade_this_learnr_ <- function(
+  label = NULL,
+  solution_code = NULL,
+  user_code = NULL,
+  check_code = NULL,
+  envir_result = NULL,
+  evaluate_result = NULL,
+  envir_prep = NULL,
+  last_value = NULL,
+  ...
+) {
+
+  learnr_args <- list(
+    label = label,
+    solution_code = solution_code,
+    user_code = user_code,
+    check_code = check_code,
+    envir_result = envir_result,
+    evaluate_result = evaluate_result,
+    envir_prep = envir_prep,
+    last_value = last_value,
+    ...
+  )
+
+  user_code <- tryCatch(
+    parse(text = user_code %||% ""),
+    error = function(e) {
+      parse_checker <- getOption(
+        "exercise.parse.error",
+        function(...) {
+          fail(paste(
+            "Uh oh, the R code produced a syntax error:",
+            conditionMessage(e),
+            "\nCheck that you have closed every \", ', (, and { ",
+            "with a matching \", ', ), and }. Also look for missing ",
+            "commas. R cannot determine how to turn your text into ",
+            "a complete command."
+          ))
+        }
+      )
+      # TODO? check that parse_checker is a function with proper args
+      eval_gradethis_expr({
+        do.call(parse_checker, append(list(parse_error = e), learnr_args))
+      })
+    }
+  )
+
+  if (is_graded(user_code)) {
+    user_code <- feedback(user_code)
+  }
+  if (is_feedback(user_code)) {
+    return(user_code)
+  }
+  if (length(user_code) == 0) {
+    return(feedback(
+      fail("I didn't receive your code. Did you write any?"),
+      type = "info"
+    ))
+  }
+
+  # Sometimes no solution is provided, but that
+  # means there is nothing to check against. Also,
+  # you do not want to parse NULL
+  if (!is.null(solution_code)) {
```

Comment:
Move this check to when solution is actually touched (so it doesn't interfere when the solution is not needed)